app/models/katello/concerns/redhat_extensions.rb

```diff
@@ -17,10 +17,13 @@ def find_or_create_operating_system(distribution)
           major, minor = distribution.version.split('.')
           minor ||= '' # treat minor versions as empty string to not confuse with nil
 
-          os = ::Redhat.where(:name => os_name, :major => major, :minor => minor).first
-          os = create_operating_system(os_name, major, minor) unless os
+          create_os = lambda { ::Redhat.where(:name => os_name, :major => major, :minor => minor).first_or_create! }
 
-          return os
+          begin
+            create_os.call
+          rescue ActiveRecord::RecordInvalid
+            create_os.call
```

Comment:
I could just call the where part again, if you prefer that I can change it:

```
begin
  ::Redhat.where(:name => os_name, :major => major, :minor => minor).first_or_create!
rescue
  ::Redhat.where(:name => os_name, :major => major, :minor => minor)
end
```

the OS is returned from the function as the begin/rescue clause is the last statement.  If you'd prefer to save it to a variable I can, but we typically don't for if/else's at the end of a function i don't think?